### ISSUE TITLE
tcplife

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bcc#693]
This tool idea came up in #690 with @iamkafai, and would print a line of output each time a TCP sessions ends, including details such as the duration of the session, and throughput statistics (from tcp_info). It could look like this:

```
# ./tcplife
PID    COMM         SADDR,PORT                      DADDR,PORT                      RX_KB  TX_KB  DUR(ms)
12345  wget         100.153.223.157,12345           123.23.100.231,80                1182      2    10.65
12345  wget         fe80::8a31:9dff:fed5:6b19,12345 fe80::8a31:9dff:fed5:6b19,80000  1182      2    10.65
```

Or put the addr/ports last to shorten the width (can fit in 80 chars for IPv4 this way):

```
# ./tcplife
PID    COMM       DUR(ms)  RX_KB  TX_KB  SADDR,PORT            DADDR,PORT
12345  wget        813.31 103871      3  100.153.223.157,12345 123.23.100.231,80
12345  wget         11.01   4713    183  fe80::8a31:9dff:fed5:6b19,12345 fe80::8a31:9dff:fed5:6b19,80
```

Or something else entirely, like ditching columns altogether:

```
# ./tcplife
TIME PID COMM SADDR,PORT -> DADDR,PORT DUR(ms) RX_KB TX_KB
0.748172 12345 wget 100.153.223.157,12345 -> 123.23.100.231,80 813.31 103871 3
```

Or use of key=values:

```
# ./tcplife
0.748172 12345 wget 100.153.223.157,12345 -> 123.23.100.231,80 ms=813.31 RX_KB=103871 TX_KB=3
```

Other details:

- DUR(ms) is the time from socket creation (or the first tcp_set_state() event seen) to when it's closed.
- -t to add a 1st "TIME" column of HH:MM:SS
- -T to add a 1st "TIME(s)" column of seconds.microseconds
- -P PORT: to filter on a destination port
- -p PID: PID filter
- -v to add trailing verbose columns: in/out segs from tcp_info

**Comment by brendangregg** [comment]
Making progress:

```
# ./tcplife.py
PID    COMM         IP SADDR            DADDR            DPORT TX_KB RX_KB MS
18785  recordProgra 4  127.0.0.1        127.0.0.1        29690     0     0 0.23
18825  recordProgra 4  127.0.0.1        127.0.0.1        28527     0     0 0.23
18825  recordProgra 4  127.0.0.1        127.0.0.1        29692     0     0 0.24
18751  sshd         4  100.66.3.172     100.127.64.230   58385     5 10255 2404.76
18853  recordProgra 4  127.0.0.1        127.0.0.1        28527     0     0 0.22
18853  recordProgra 4  127.0.0.1        127.0.0.1        29694     0     0 0.23
18876  recordProgra 4  127.0.0.1        127.0.0.1        28527     0     0 0.26
```

This is just one kprobe: tcp_set_state(), and I'm getting lifespan and TX/RX Kbytes. :)

I need to do something about the width though...

**Comment by brendangregg** [comment]
Here's my new strategy for fetching task context (best effort):

- active: save state on TCP_SYN_SENT
- passive: read state on TCP_LAST_ACK

tcplife.py test:

```
PID   COMM       LADDR           LPORT RADDR           RPORT TX_KB RX_KB MS
11627 ssh        127.0.0.1       31578 127.0.0.1       22        1     1 169.56
11628 sshd       127.0.0.1       22    127.0.0.1       31578     1     1 171.26

11770 curl       100.66.3.172    63188 209.85.232.147  80        0    11 36.71

11846 sshd       100.66.3.172    22    100.127.64.230  3627      3     3 1378.69
```

The first pair is a localhost connection.
The curl is an active connection.
The sshd is a passive connection.
All work.

**Comment by brendangregg** [comment]
fixed by #773.
